stats/csv/collector.go

```diff
@@ -0,0 +1,231 @@
+		}, nil
+	}
+
+	logfile, err := fs.Create(fname)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Collector{
+		outfile:      logfile,
+		fname:        fname,
+		resTags:      resTags,
+		ignoredTags:  ignoredTags,
+		csvWriter:    csv.NewWriter(logfile),
+		row:          make([]string, 0, 3+len(resTags)+1),
+		saveInterval: saveInterval,
+	}, nil
+}
+
+// Init writes column names to csv file
+func (c *Collector) Init() error {
+	header := MakeHeader(c.resTags)
+	err := c.csvWriter.Write(header)
+	if err != nil {
+		logrus.WithField("filename", c.fname).Error("CSV: Error writing column names to file")
+	}
+	c.csvWriter.Flush()
+	return nil
+}
+
+// SetRunStatus does nothing
+func (c *Collector) SetRunStatus(status lib.RunStatus) {}
+
+// Run just blocks until the context is done
+func (c *Collector) Run(ctx context.Context) {
+	ticker := time.NewTicker(c.saveInterval)
+	for {
+		select {
+		case <-ticker.C:
+			c.WriteToFile()
+		case <-ctx.Done():
+			c.WriteToFile()
+			err := c.outfile.Close()
+			if err != nil {
+				logrus.WithField("filename", c.fname).Error("CSV: Error closing the file")
+			}
+			return
+		}
+	}
+}
+
+// Collect Saves samples to buffer
+func (c *Collector) Collect(scs []stats.SampleContainer) {
+	c.bufferLock.Lock()
+	defer c.bufferLock.Unlock()
+	for _, sc := range scs {
+		c.buffer = append(c.buffer, sc.GetSamples()...)
+	}
+}
+
+// WriteToFile Writes samples to the csv file
+func (c *Collector) WriteToFile() {
+	c.bufferLock.Lock()
+	samples := c.buffer
+	c.buffer = nil
+	c.bufferLock.Unlock()
+
+	if len(samples) > 0 {
+		c.csvLock.Lock()
+		defer c.csvLock.Unlock()
+		for _, sc := range samples {
+			for _, sample := range sc.GetSamples() {
+				sample := sample
+				row := c.row[:0]
+				row = SampleToRow(&sample, c.resTags, c.ignoredTags, row)
+				err := c.csvWriter.Write(row)
+				if err != nil {
+					logrus.WithField("filename", c.fname).Error("CSV: Error writing to file")
+				}
+			}
+		}
+		c.csvWriter.Flush()
+	}
+}
+
+// Link returns a dummy string, it's only included to satisfy the lib.Collector interface
+func (c *Collector) Link() string {
+	return c.fname
+}
+
+// MakeHeader creates list of column names for csv file
+func MakeHeader(tags []string) []string {
+	tags = append(tags, "extra_tags")
+	return append([]string{"metric_name", "timestamp", "metric_value"}, tags...)
+}
+
+// SampleToRow converts sample into array of strings
+func SampleToRow(sample *stats.Sample, resTags []string, ignoredTags []string, row []string) []string {
+	row = append(row, sample.Metric.Name)
+	row = append(row, fmt.Sprintf("%d", sample.Time.Unix()))
+	row = append(row, fmt.Sprintf("%f", sample.Value))
+	sampleTags := sample.Tags.CloneTags()
+
+	for _, tag := range resTags {
+		row = append(row, sampleTags[tag])
+	}
+
+	extraTags := bytes.Buffer{}
+	prev := false
+	for tag, val := range sampleTags {
+		if sort.SearchStrings(resTags, tag) == len(resTags) && sort.SearchStrings(ignoredTags, tag) == len(ignoredTags) {
```

Comment:
As seen in the [sort.Search](https://golang.org/pkg/sort/#Search) documentation and the example below in particular, Search* returns where the element should be added. So while returning the length of the slice is good indicator that the element is not inside of the slice it could return an index inside of the slice and still not be present. I probably should've added this before, sorry. 

Ideally you will first add a test that fails in the current situation and than fix it ;). 

p.s. I recommend adding a helper function that takes a (sorted) slice and and element and returns whether the element is inside ;)